README.md
# i-doit CLI tool

Access your CMDB on the command line interface

[![Latest Stable Version](https://img.shields.io/packagist/v/bheisig/idoitcli.svg)](https://packagist.org/packages/bheisig/idoitcli)
[![Minimum PHP Version](https://img.shields.io/badge/php-%3E%3D%207.0-8892BF.svg)](https://php.net/)
[![Build Status](https://travis-ci.org/bheisig/i-doit-cli.svg?branch=master)](https://travis-ci.org/bheisig/i-doit-cli)

## About

[i-doit](https://i-doit.com) is a software application for IT documentation and a CMDB (Configuration Management Database). This application is very useful to collect all your knowledge about the IT infrastructure you are dealing with. i-doit is a Web application and [has an exhausting API](https://kb.i-doit.com/pages/viewpage.action?pageId=37355644) which is very useful to automate your infrastructure.

This application provides a simple, but powerful command line interface to access your CMDB data stored in i-doit.

## Features

*   Read information about objects their types and even their attributes
*   Find your needle in the haystack called CMDB
*   Do you need a free IP address in a particular subnet? This app suggests one for you.
*   Stress your system: auto-generate thousands of objects

## Requirements

Before using this app your system must meet the following requirements:

*   Of course, i-doit pro/open, version 1.13 or higher
*   i-doit API add-on, version 1.10.3 or higher
*   Any POSIX operating system (GNU/Linux, *BSD, MacOS) or Windows
*   PHP >= 7.1 (7.3 is recommended)
*   PHP extensions `calendar`, `cli`, `cURL`, `json`, `phar` and `zlib`
*   PHP extension `pcntl` is optional, but highly recommended (non-Windows only)

## Install and update

You have two good options to download and install this application:

*   Download any stable release (**recommended**)
*   Use [PHIVE](https://phar.io/)

### Download release

Download the latest stable version of the binary `idoitcli` [from the release site](https://github.com/bheisig/i-doit-cli/releases). Then install it system-wide:

~~~ {.bash}
curl -OL https://github.com/bheisig/i-doit-cli/releases/download/0.7/idoitcli
chmod 755 idoitcli
sudo mv idoitcli /usr/local/bin/
~~~

To be up-to-date just repeat the steps above.

### Use PHIVE

With [PHIVE](https://phar.io/) you are able to download and install PHAR files on your system. Additionally, it will verify the SHA1 and GPG signatures which is highly recommended. If you have PHIVE already installed you can fetch the latest version of this application:

~~~ {.bash}
sudo phive install --global bheisig/i-doit-cli
~~~

This will install an executable binary to `/usr/bin/idoitcli`.

If a new release is available you can perform an update:

~~~ {.bash}
sudo phive update --global
~~~

## Usage

Just run the application to show the basic usage:

~~~ {.bash}
idoitcli
~~~

## First steps

This application caches a lot locally to give you the best user experience. Run the `init` command:

~~~ {.bash}
idoitcli init
~~~

Some simple questions will be asked how to access your i-doit installation. Next step is to create cache files:

~~~ {.bash}
idoitcli cache
~~~

After that some files will be created in your home directory under `~/.idoitcli/`: Each i-doit installation has its own cache files under `data/`. Your user-defined configuration file is called `config.json`.

You may check your current status by running:

~~~ {.bash}
idoitcli status
~~~

This gives you some basic information about your i-doit installation, your settings and your user.

## Access your CMDB data

This is probably the best part: Read information about objects, their types and even attributes.

List all object types:

~~~ {.bash}
idoitcli read
idoitcli read /
~~~

List server objects:

~~~ {.bash}
idoitcli read server/
idoitcli read server/host.example.net
idoitcli read server/*.example.net
idoitcli read server/host.*.net
idoitcli read server/*.*.net
idoitcli read server/host*
idoitcli read server/*
~~~

Show common information about server "host.example.net":

~~~ {.bash}
idoitcli read server/host.example.net
~~~

Show common information about object identifier "42":

~~~ {.bash}
idoitcli read 42
~~~

Show common information about one or more objects by their titles:

~~~ {.bash}
idoitcli read *.example.net
idoitcli read host.*.net
idoitcli read *.*.net
idoitcli read host*
~~~

List assigned categories:

~~~ {.bash}
idoitcli read server/host.example.net/
~~~

Show values from category "model" for this server:

~~~ {.bash}
idoitcli read server/host.example.net/model
~~~

Show values from category "model" for one or more servers:

~~~ {.bash}
idoitcli read server/*.example.net/model
idoitcli read server/host.*.net/model
idoitcli read server/*.*.net/model
idoitcli read server/host*/model
idoitcli read server/*/model
~~~

Or just show the name of the model:

~~~ {.bash}
idoitcli read server/host.example.net/model/model
~~~

List available attributes for category "model":

~~~ {.bash}
idoitcli read server/host.example.net/model/
~~~

You may leave the object type empty for specific objects, for example:

~~~ {.bash}
idoitcli read host.example.net/model
~~~

**Notice:** These examples work great with unique names. That is why it is common practice to give objects unique titles that are not in conflict with object types and categories.

## Show everything about an object

~~~ {.bash}
idoitcli show myserver
idoitcli show "My Server"
idoitcli show 42
~~~

## Find your data

Find your needle in the haystack called CMDB:

~~~ {.bash}
idoitcli search myserver
idoitcli search "My Server"
~~~

## Show the next free IPv4 address

Get the next free IPv4 address for a particular subnet:

~~~ {.bash}
idoitcli nextip SUBNET
~~~

`SUBNET` may be the object title or its identifier.

## Auto-generate objects

For testing purposes stress your i-doit installation and let the app create thousands of objects, attributes and relations between objects:

~~~ {.bash}
idoitcli -c FILE random
~~~

There are some examples located under `docs/`.

## Update the caches

If your CMDB configuration has changed you need to re-create the cache files by running the `cache` command:

~~~ {.bash}
idoitcli cache
~~~

## Playground

Perform self-defined API requests – pass request as argument:

~~~ {.bash}
idoitcli call '{"version": "2.0","method": "idoit.version","params": {"apikey": "c1ia5q","language": "en"},"id": 1}'
~~~

Pipe request:

~~~ {.bash}
echo '{"version": "2.0","method": "idoit.version","params": {"apikey": "c1ia5q","language": "en"},"id": 1}' | idoitcli call
~~~

Read request from file:

~~~ {.bash}
cat request.txt | idoitcli call
~~~

Read request from standard input (double-enter to execute):

~~~ {.bash}
idoitcli call
~~~

## Configuration

There are some ways to set your configurations settings:

1.  System-wide settings are stored under `/etc/idoitcli/config.json`.
2.  User-defined settings are stored under `~/.idoitcli/config.json`.
3.  Pass your run-time settings by using the options `-c` or `--config`.

Keep in mind this order matters. Each file is optional. Any combination is possible. Furthermore, to include even more configuration files you can pass the options `-c` and `--config` as often as you like.

The configuration files are JSON-formatted.

## Contribute

Please, report any issues to [our issue tracker](https://github.com/bheisig/i-doit-cli/issues). Pull requests and OS distribution packages are very welcomed. For further information, see file [`CONTRIBUTING.md`](CONTRIBUTING.md).

## Copyright & License

Copyright (C) 2016-19 [Benjamin Heisig](https://benjamin.heisig.name/)

Licensed under the [GNU Affero GPL version 3 or later (AGPLv3+)](https://gnu.org/licenses/agpl.html). This is free software: you are free to change and redistribute it. There is NO WARRANTY, to the extent permitted by law.

[List of first names and gender:](https://www.heise.de/ct/ftp/07/17/182/) Copyright (C) 2007-2008 Jörg Michael, licensed under [GNU Free Documentation License](https://www.gnu.org/licenses/fdl.html)

[List of surnames:](https://github.com/HBehrens/phonet4n/blob/master/src/Tests/data/nachnamen.txt) Copyright (C) Heiko Behrens, lisenced under [GNU Lesser General Public License](https://www.gnu.org/licenses/lgpl-3.0)


CHANGELOG.md
#   Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased][]

_tbd_

## [0.9][] – 2020-01-08

### Added

-   `network`: Add new command to print list of IPv4 addresses
-   `rack`: Add new command to visualize hardware racks

### Changed

-   `save`: Handle more attributes as text fields

### Fixed

-   `save`: Do not ignore capacity (float) incl. unit (b, kb, mb, gb, tb)
-   `read`: Print object types when no arguments are given

##  [0.8][] – 2019-07-31

☀️☀️☀️ Happy summer time ☀️☀️☀️

##  Changed

-   Require PHP >= 7.1 (PHP 7.0 is end-of-life); recommend PHP 7.3
-   i-doit 1.13 or higher is required
-   i-doit API add-on 1.10.3 is required
-   Require PHP extension `zlib` instead of `bz2`

## Fixed

-   Allow API calls without credentials
-   `save`/`show`: Prevent PHP fatal error when encoding empty dialog+ attribute
-   `show`: Prevent any unwanted user interaction (`--yes`) if object not found or selection ambiguous
-   Make CLI option `-y`/`--yes` work after it was completely ignored
-   `read`: Convert values to strings before printing them which prevents PHP fatal errors

## [0.7][] – 2018-12-18

This release is a great step forward with new features and tons of improvements. Happy holidays!

### Added

-   `save`: Create/update CMDB objects and their category entries
-   `log`: Add entry to i-doit logbook
-   `logs`: Print entries from i-doit logbook
-   `cache`: Create cache files needed for faster processing
-   Add support for custom categories with user-defined attributes

### Changed

-   `init`: Just create configuration file and do not cache files anymore
-   `help`/`-h`/`--help`: Print more information about using each command
-   Ignore categories which have no proper attributes

### Fixed

-   `init`: Ask to enable proxy settings

## [0.6][]

### Added

-   Support for Windows operating systems
-   `random`: Create persons with random names, e-mail addresses and desks
-   `random`: Install applications on laptops with license keys

### Fixed

-   `init`: Validation error occurs even before initialization could start
-   `init`: Ask for hostname if HTTP proxy will be used

## [0.5][] – 2018-04-24

**Important notes:**

1.  PHP 5.6 support is dropped. You need PHP 7.0 or higher.
2.  Re-load cache with `idoitcli init`

### Added

-   `create`: Create new object or category entry

### Changed

-   Renamed binary file to `idoitcli`
-   Separate CLI-related code from application code
-   Drop PHP 5.6 support and require PHP 7.0 or higher

### Fixed

-   `show`: Not all category entries are printed
-   `init`: Misleading cache file names for categories

## [0.4][] – 2017-09-21

### Added

-   `read`: Allow wildcards (`*`) in object titles
-   `categories`: Print a list of available categories
-   `types`: Print a list of available object types and group them
-   `fixip`: Assign IPs to subnets
-   Display error when user asks for specific category entries but category is not assigned to object type(s)

## [0.3][] – 2017-07-25

### Added

-   `show`: Show everything about an object
-   `read`: Read object by its identifier
-   `call`: Perform self-defined API requests
-   Find errors in additional configuration files

### Fixed

-   Show the right command description (`idoit help [command]` or `idoit [command] --help`)
-   Command `idoit read` was unable to fetch a list of objects or attributes for an object when configuration setting `limitBatchRequest` is disabled (`0`).

## [0.2][] – 2017-04-06

### Added

-   Find your data with new command `search`
-   List attributes and assigned categories with command `read`
-   Sort results with command `read`
-   Strip HTML code in results for command `read`
-   Limit batch requests (if configured) for command `read` with setting `limitBatchRequests`
-   Describe a lot of examples for command `read` in [documentation](README.md) und built-in help
-   Simple bash completion
-   Create racks and servers with categories "formfactor", "cpu", "model", and "location" with command `random`
-   Put servers into empty racks with command `random`
-   More examples in `docs/` for command `random`
-   Show message if cache is out-dated, see configuration setting `cacheLifetime`

### Fixed

-   Errors in built-in `help`
-   Removed out-dated command line options
-   Source code documentation

## 0.1 – 2017-02-06

Initial release

[Unreleased]: https://github.com/bheisig/i-doit-cli/compare/0.9...HEAD
[0.9]: https://github.com/bheisig/i-doit-cli/compare/0.8...0.9
[0.8]: https://github.com/bheisig/i-doit-cli/compare/0.7...0.8
[0.7]: https://github.com/bheisig/i-doit-cli/compare/0.6...0.7
[0.6]: https://github.com/bheisig/i-doit-cli/compare/0.5...0.6
[0.5]: https://github.com/bheisig/i-doit-cli/compare/0.4...0.5
[0.4]: https://github.com/bheisig/i-doit-cli/compare/0.3...0.4
[0.3]: https://github.com/bheisig/i-doit-cli/compare/0.2...0.3
[0.2]: https://github.com/bheisig/i-doit-cli/compare/0.1...0.2


src/Call.php
<?php

/**
 * Copyright (C) 2017 Benjamin Heisig
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 *
 * @author Benjamin Heisig <https://benjamin.heisig.name/>
 * @copyright Copyright (C) 2016-17 Benjamin Heisig
 * @license http://www.gnu.org/licenses/agpl-3.0 GNU Affero General Public License (AGPL)
 * @link https://github.com/bheisig/i-doit-cli
 */

namespace bheisig\idoitcli;

/**
 * Command "call"
 */
class Call extends Command {

    use APICall;

    /**
     * Processes some routines before the execution
     *
     * @return self Returns itself
     *
     * @throws \Exception on error
     */
    public function setup() {
        parent::setup();

        $this->initiateAPI();

        return $this;
    }

    /**
     * Executes the command
     *
     * @return self Returns itself
     *
     * @throws \Exception on error
     */
    public function execute() {
        $query = $this->getQuery();

        if ($query === '') {
            while (true) {
                $line = IO::in('');

                if ($line === '') {
                    break;
                }

                $query .= $line;
            }
        }

        $request = json_decode(trim($query), true);

        if (!is_array($request)) {
            throw new \Exception('No valid JSON input');
        }

        if (!array_key_exists('method', $request)) {
            throw new \Exception('No RPC method specified');
        }

        $params = null;

        if (array_key_exists('params', $request)) {
            $params = $request['params'];
        }

        $this->api->request($request['method'], $params);

        IO::out('');
        IO::out('Request:');
        IO::out('========');
        IO::out($this->api->getLastRequestHeaders());
        IO::out(json_encode($this->api->getLastRequestContent(), JSON_PRETTY_PRINT));
        IO::out('');
        IO::out('Response:');
        IO::out('=========');
        IO::out($this->api->getLastResponseHeaders());
        IO::out(json_encode($this->api->getLastResponse(), JSON_PRETTY_PRINT));

        return $this;
    }

    /**
     * Processes some routines after the execution
     *
     * @return self Returns itself
     *
     * @throws \Exception on error
     */
    public function tearDown () {
        return parent::tearDown();
    }

    /**
     * Shows usage of this command
     *
     * @return self Returns itself
     */
    public function showUsage() {
        $command = strtolower((new \ReflectionClass($this))->getShortName());

        IO::out('Usage: %1$s [OPTIONS] %2$s [REQUEST]

%3$s

REQUEST is a JSON-formatted string. Leave empty to read from standard input.

Examples:

1) %1$s %2$s \'{"version": "2.0","method": "idoit.version","params": {"apikey": "c1ia5q","language": "en"},"id": 1}\'
2) echo \'{"version": "2.0","method": "idoit.version","params": {"apikey": "c1ia5q","language": "en"},"id": 1}\' | %1$s %2$s
3) cat request.txt | %1$s %2$s',
            $this->config['basename'],
            $command,
            $this->config['commands'][$command]
        );

        return $this;
    }

}
